primitives/src/balances.rs

```diff
@@ -56,6 +57,92 @@ pub struct BalanceEntry<BlockNumber> {
 	pub last_update: BlockNumber,
 }
 
+impl<BlockNumber> BalanceEntry<BlockNumber>
+where
+	BlockNumber: AtLeast32Bit,
+{
+	pub fn new(principal: BalanceType, last_update: BlockNumber) -> Self {
+		Self { principal, last_update }
+	}
+
+	/// Applies the demurrage and returns an updated BalanceEntry.
+	///
+	/// The following formula is applied to the principal:
+	///    updated_principal = old_principal * e^(-demurrage_per_block * elapsed_blocks)
+	///
+	/// **Note**: This function will be used at every single transaction that is paid with community
+	/// currency. It is important that it is as efficient as possible, but also that it is bullet-
+	/// proof (no-hidden panics!!).
+	pub fn apply_demurrage(
+		self,
+		demurrage_per_block: Demurrage,
+		current_block: BlockNumber,
+	) -> Result<BalanceEntry<BlockNumber>, DemurrageError> {
+		if demurrage_per_block < 0 {
+			return Err(DemurrageError::DemurrageMustBePositive)
+		}
+
+		if self.last_update == current_block {
+			// Nothing to be done, as no time elapsed.
+			return Ok(self)
+		}
+
+		if self.principal.eq(&0i16) {
+			return Ok(Self { principal: self.principal, last_update: current_block })
+		}
+
+		let elapsed_blocks = current_block
+			.checked_sub(&self.last_update)
+			.ok_or(DemurrageError::LastBlockBiggerThanCurrent)?;
+
+		let elapsed_u32: u32 = elapsed_blocks
+			.try_into()
+			.map_err(|_| DemurrageError::ElapsedBlocksMoreThan32Bits)?; // 698 years with 6 seconds block time.
+
+		let demurrage_factor = demurrage_factor(demurrage_per_block, elapsed_u32)?;
+
+		let principal = self
+			.principal
+			.checked_mul(demurrage_factor)
+			.expect("demurrage_factor [0,1), hence can't overflow; qed");
+
+		Ok(Self { principal, last_update: current_block })
+	}
+}
+
+/// e^(-demurrage_per_block * elapsed_blocks) within [0,1).
+///
+/// Returns an error if the demurrage is negative.
+pub fn demurrage_factor(
+	demurrage_per_block: Demurrage,
+	elapsed_blocks: u32,
+) -> Result<BalanceType, DemurrageError> {
+	if demurrage_per_block < 0 {
+		return Err(DemurrageError::DemurrageMustBePositive)
+	}
+
+	// We can only have errors here if one of the operations overflowed. However, we take the
+	// inverse of these operations at the end, which is why we can set the demurrage factor to
```

Comment:
"inverse" only makes sense if one knows about what the `exp` function does. There is no inverse in this function.

How about:
> However, if the calculation overflows, the exponent is too big to calculate. In this case (because e^(-x) -> 0 for big x) we can approximate to 0.

Reply:
Thanks a lot, I think you are right, I have updated the documentation accordingly, I hope it is better now.